### ISSUE TITLE
Distributive property of matmul is violated

### ISSUE THREAD

**Issue by kristian-georgiev** [pytorch/pytorch#94628]
### 🐛 Describe the bug

Matrix multiplication does not obey the distributive property with respect to (matrix) addition, i.e. $A (B + C) \neq AB + AC$ for `torch.matmul`. The behavior can be observed on both CPU and GPU, both for `float16` and `float32` tensors.
For example, the following code
```
import torch
for device, dtype in [('cuda:0', torch.float16), ('cuda:0', torch.float32), ('cpu', torch.float32)]:
        A = torch.randn(10_000, 10_000, dtype=dtype).to(device)
        B = torch.randn(10_000, dtype=dtype).to(device)
        C = torch.randn(10_000, dtype=dtype).to(device)

        result_1 = A @ B + A @ C
        result_2 = A @ (B + C)

        print(f'Device: {device}, dtype: {dtype}')
        print(f'Are all values close: {torch.allclose(result_1, result_2)}')
        print(f'The norm of the difference is: {(result_1 - result_2).norm()}')
        print('=' * 20)
```
produces
```
Device: cuda:0, dtype: torch.float16
Are all values close: False
The norm of the difference is: 6.09765625
====================
Device: cuda:0, dtype: torch.float32
Are all values close: False
The norm of the difference is: 0.009012020193040371
====================
Device: cpu, dtype: torch.float32
Are all values close: False
The norm of the difference is: 0.018055615946650505
```
(`float16` + `cpu` not listed since `addmv_impl_cpu` is not implemented for `float16`)

In comparison, `numpy` maintains dramatically better numerical precision. The following code
```
A = np.random.randn(10_000, 10_000)
B = np.random.randn(10_000)
C = np.random.randn(10_000)

result_1 = A @ B + A @ C
result_2 = A @ (B + C)

print(f'Are all values close: {np.allclose(result_1, result_2)}')
print(f'The norm of the difference is: {np.linalg.norm(result_1 - result_2)}')
```
produces
```
Are all values close: True
The norm of the difference is: 1.5367062820716133e-11
```

### Versions

Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (conda-forge gcc 11.3.0-19) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.27

Python version: 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 15:55:03)  [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-4.15.0-193-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.3.58
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-PCIE-40GB
GPU 1: NVIDIA A100-PCIE-40GB
GPU 2: NVIDIA A100-PCIE-40GB
GPU 3: NVIDIA A100-PCIE-40GB
GPU 4: NVIDIA A100-PCIE-40GB
GPU 5: NVIDIA A100-PCIE-40GB
GPU 6: NVIDIA A100-PCIE-40GB
GPU 7: NVIDIA A100-PCIE-40GB
GPU 8: NVIDIA A100-PCIE-40GB

Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7402 24-Core Processor
Stepping:            0
CPU MHz:             1499.454
CPU max MHz:         2800.0000
CPU min MHz:         1500.0000
BogoMIPS:            5599.90
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-23,48-71
NUMA node1 CPU(s):   24-47,72-95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate sme ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] open-clip-torch==2.9.3
[pip3] pytorch-pfn-extras==0.6.3
[pip3] torch==1.13.0.post200
[pip3] torchvision==0.14.0a0+2ba5a5d
[conda] cudatoolkit               11.3.1              h9edb442_11    conda-forge
[conda] magma                     2.6.2                hc72dce7_0    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] numpy                     1.23.5           py39h3d75532_0    conda-forge
[conda] open-clip-torch           2.9.3                    pypi_0    pypi
[conda] pytorch                   1.13.0          cuda112py39hb0b7ed5_200    conda-forge
[conda] pytorch-pfn-extras        0.6.3                    pypi_0    pypi
[conda] torchvision               0.14.0          cuda112py39h1a1de93_0    conda-forge

**Comment by ngimel** [comment]
This is expected with large matrices and limited floating point accuracy https://pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by GuillaumeLeclerc** [comment]
@ngimel Isn't numpy also using IEEE 754 ? How do you explain 9 orders of magnitude difference in the error ?

**Comment by GuillaumeLeclerc** [comment]
Any update @ngimel ? It seems reasonable to expect two library using the same floating point representation to not have 9 orders of magnitude difference in accuracy no ?

**Comment by albanD** [comment]
> Isn't numpy also using IEEE 754 ?

The difference is that numpy uses double precision floating points by default while pytorch use single precision floating points. So yes 9 orders of magnitude is exactly the difference we expect.
